File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/ask_user.py
# nvdaMcpBridge domain -- AskUserHandler: present a prompt to the human.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: command handler for `askUser`; returns the ticket at once, and the agent polls waitForUserReply.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from ...entities.user_prompt import UserPrompt
from .command_handler import CommandError, CommandHandler

if TYPE_CHECKING:
	from .session_context import SessionContext


class AskUserHandler(CommandHandler):
	mutates_reader = True

	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		params = protocol.from_dict(protocol.AskUserParams, request.params)

		prompt = UserPrompt(params.prompt, ctx.clock)
		if not ctx.set_outstanding_prompt(prompt):
			raise CommandError(
				"a prompt is already outstanding; wait for it or let it expire before asking another"
			)

		ctx.suspend_speech()
		ctx.user_prompter.present(params.prompt, prompt.ticket)
		# Reset the silence cap after the prompt is presented, not before.
		ctx.note_audible()

		ctx.transcript.note(f"askUser: prompt presented (ticket {prompt.ticket})")
		return protocol.AskUserResult(ticket=prompt.ticket)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/AskUser.swift
// ROLE: controller for `askUser`: puts a question to the human and returns a ticket to
// collect the answer with.
// BUILT BY: Registry. DRIVES: the UserPrompter, Announcer and SilenceControl ports, and the
// session's silence cap.
// Returns at once: the session thread renews the silence lease, so waiting here for the human
// would unmute the machine.
// A silent session passes through while the window is open; `suspendedSilence` records that,
// so `waitForUserReply` restores only what this took.

import Foundation
import ScreenReaderWire

public final class AskUserHandler: CommandHandler {
	public let mutatesReader = true

	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let params = try request.params(as: AskUserParams.self)
		let question = params.prompt.trimmingCharacters(in: .whitespacesAndNewlines)
		guard !question.isEmpty else {
			throw CommandError("`askUser` needs a prompt: an empty question cannot be answered")
		}
		guard let adapters = context.adapters else {
			throw CommandError("`askUser` was called before `hello` built the reader edge")
		}
		guard context.outstandingPrompt == nil else {
			throw CommandError(
				"a prompt is already outstanding on this session; collect it with `waitForUserReply` "
					+ "or let its window expire before asking another")
		}

		let ticket: PromptId
		do {
			ticket = try adapters.userPrompter.present(question)
		} catch {
			throw CommandError("the question could not be put to the human: \(describe(error))")
		}
		let prompt = UserPrompt(ticket: ticket, prompt: question, now: context.clock.monotonic())

		if context.mode == .silent, adapters.silenceControl.isSuppressing {
			do {
				try adapters.silenceControl.passThrough()
				prompt.suspendedSilence = true
				context.transcript.note("askUser: silence suspended while the prompt is open")
			} catch {
				context.transcript.note(
					"askUser: the reader could not be un-muted for the prompt: \(describe(error))")
			}
		}
		context.outstandingPrompt = prompt

		do {
			try adapters.announcer.announce(question)
		} catch {
			// The window is up and the ticket is real, so the command still succeeds.
			context.transcript.note("askUser: the question could not be spoken: \(describe(error))")
		}
		context.transcript.note("askUser: prompt presented (ticket \(ticket))")
		context.humanHeard()
		return AskUserResult(ticket: ticket)
	}

	private func describe(_ error: any Error) -> String {
		String(describing: error)
	}
}
